Clippy slice panic on --fix

### Summary

Getting a panic message ``slice index starts at 1 but ends at 0`` by running the following command:
```
cargo +nightly-2022-05-21 clippy --fix --no-default-features
```

Which results in this stack backtrace:

<details>
    <summary>Backtrace (Click to Expand)</summary>

```
stack backtrace:
   0:     0x7ff6dd68b3bf - git_midx_writer_dump
   1:     0x7ff6dd6b112a - git_midx_writer_dump
   2:     0x7ff6dd6818e9 - git_midx_writer_dump
   3:     0x7ff6dd68e67b - git_midx_writer_dump
   4:     0x7ff6dd68e26e - git_midx_writer_dump
   5:     0x7ff6dd68ec71 - git_midx_writer_dump
   6:     0x7ff6dd68eb2d - git_midx_writer_dump
   7:     0x7ff6dd68bf77 - git_midx_writer_dump
   8:     0x7ff6dd68e809 - git_midx_writer_dump
   9:     0x7ff6dd7449e5 - git_midx_writer_dump
  10:     0x7ff6dd6b2f2b - git_midx_writer_dump
  11:     0x7ff6dd6a9659 - git_midx_writer_dump
  12:     0x7ff6dd6ae4f9 - git_midx_writer_dump
  13:     0x7ff6dd744b29 - git_midx_writer_dump
  14:     0x7ff6dd519b48 - git_midx_writer_dump
  15:     0x7ff6dd519e8d - git_midx_writer_dump
  16:     0x7ff6dd29efdf - git_filter_source_path
  17:     0x7ff6dd29f7a2 - git_filter_source_path
  18:     0x7ff6dcd957e9 - <unknown>
  19:     0x7ff6dd019435 - git_odb_object_id
  20:     0x7ff6dcd0c270 - <unknown>
  21:     0x7ff6dcd562d6 - <unknown>
  22:     0x7ff6dcd7d50c - <unknown>
  23:     0x7ff6dd678411 - git_midx_writer_dump
  24:     0x7ff6dcd0e877 - <unknown>
  25:     0x7ff6dd6b96d0 - git_midx_writer_dump
  26:     0x7fffbbd9244d - BaseThreadInitThunk
  27:     0x7fffbcd0df88 - RtlUserThreadStart
error: could not compile `arcropolis
```
</details>

### Reproducer

[Code and commit](https://github.com/Raytwo/ARCropolis/tree/44b16445fd9d01a63a67f5ef0252b6c98f68de71)

Simply run the following command: ``cargo +nightly-2022-05-21 clippy --fix --no-default-features``

### Version

```text
rustc 1.63.0-nightly (e6a4afc3a 2022-05-20)
binary: rustc
commit-hash: e6a4afc3af2d2a53f91fc8a77bdfe94bea375b29
commit-date: 2022-05-20
host: x86_64-pc-windows-msvc
release: 1.63.0-nightly
LLVM version: 14.0.4
```


### Additional Labels

_No response_

I can't compile your code :/ 

```text
error[E0557]: feature has been removed
  --> /home/marcel/.cargo/git/checkouts/skyline-rs-652bad0610bd03a8/3278cf5/src/lib.rs:14:13
   |
14 |     feature(const_generics, impl_trait_in_bindings, alloc_prelude)
   |             ^^^^^^^^^^^^^^ feature has been removed
   |
   = note: removed in favor of `#![feature(adt_const_params)]` and `#![feature(generic_const_exprs)]`

error[E0557]: feature has been removed
  --> /home/marcel/.cargo/git/checkouts/skyline-rs-652bad0610bd03a8/3278cf5/src/lib.rs:14:29
   |
14 |     feature(const_generics, impl_trait_in_bindings, alloc_prelude)
   |                             ^^^^^^^^^^^^^^^^^^^^^^ feature has been removed
   |
   = note: the implementation was not maintainable, the feature may get reintroduced once the current refactorings are done

error[E0433]: failed to resolve: could not find `prelude` in `alloc`
   --> /home/marcel/.cargo/git/checkouts/skyline-rs-652bad0610bd03a8/3278cf5/src/lib.rs:100:27
    |
100 |     pub use crate::alloc::prelude::v1::*;
    |                           ^^^^^^^ could not find `prelude` in `alloc`

error[E0635]: unknown feature `alloc_prelude`
  --> /home/marcel/.cargo/git/checkouts/skyline-rs-652bad0610bd03a8/3278cf5/src/lib.rs:14:53
   |
14 |     feature(const_generics, impl_trait_in_bindings, alloc_prelude)
   |                                                     ^^^^^^^^^^^^^
```

--- a/tests/ui/map_flatten_fixable.fixed
+++ b/tests/ui/map_flatten_fixable.fixed
@@ -28,4 +28,22 @@ fn main() {
 
     // mapping to Result on Result
     let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
+
+    issue8878();
+}
+
+#[allow(clippy::bind_instead_of_map)] // map + flatten will be suggested to `and_then`, but afterwards `map` is suggested again
+#[rustfmt::skip] // whitespace is important for this one
+fn issue8878() {
+    std::collections::HashMap::<u32, u32>::new()
+        .get(&0)
+        .and_then(|_| {
+// we need some newlines
+// so that the span is big enough
+// we need some newlines
+// so that the span is big enough
+// for a splitted output of the diagnostic
+            Some("")
+ // whitespace beforehand is important as well
+        });
 }

--- a/tests/ui/map_flatten_fixable.fixed
+++ b/tests/ui/map_flatten_fixable.fixed
@@ -29,7 +29,26 @@ fn main() {
     // mapping to Result on Result
     let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
 
+    issue8734();
     issue8878();
+}
+
+fn issue8734() {
+    //     let _ = [0u8, 1, 2, 3]
+    //         .into_iter()
+    //         .map(|n| match n {
+    //             1 => [n
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)],
+    //             n => [n],
+    //         })
+    //         .flatten();
 }
 
 #[allow(clippy::bind_instead_of_map)] // map + flatten will be suggested to `and_then`, but afterwards `map` is suggested again

--- a/tests/ui/map_flatten_fixable.rs
+++ b/tests/ui/map_flatten_fixable.rs
@@ -29,7 +29,26 @@ fn main() {
     // mapping to Result on Result
     let _: Result<_, &str> = (Ok(Ok(1))).map(|x| x).flatten();
 
+    issue8734();
     issue8878();
+}
+
+fn issue8734() {
+    //     let _ = [0u8, 1, 2, 3]
+    //         .into_iter()
+    //         .map(|n| match n {
+    //             1 => [n
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)
+    //                 .saturating_add(1)],
+    //             n => [n],
+    //         })
+    //         .flatten();
 }
 
 #[allow(clippy::bind_instead_of_map)] // map + flatten will be suggested to `and_then`, but afterwards `map` is suggested again

--- a/tests/ui/map_flatten_fixable.rs
+++ b/tests/ui/map_flatten_fixable.rs
@@ -28,4 +28,21 @@ fn main() {
 
     // mapping to Result on Result
     let _: Result<_, &str> = (Ok(Ok(1))).map(|x| x).flatten();
+
+    issue8878();
+}
+
+#[allow(clippy::bind_instead_of_map)] // map + flatten will be suggested to `and_then`, but afterwards `map` is suggested again
+#[rustfmt::skip] // whitespace is important for this one
+fn issue8878() {
+    std::collections::HashMap::<u32, u32>::new()
+        .get(&0)
+        .map(|_| {
+// we need some newlines
+// so that the span is big enough
+// for a splitted output of the diagnostic
+            Some("")
+ // whitespace beforehand is important as well
+        })
+        .flatten();
 }

--- a/tests/ui/map_flatten_fixable.stderr
+++ b/tests/ui/map_flatten_fixable.stderr
@@ -76,5 +76,31 @@ help: try replacing `map` with `and_then`, and remove the `.flatten()`
 LL |     let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
    |                                          ~~~~~~~~~~~~~~~
 
-error: aborting due to 7 previous errors
+error: called `map(..).flatten()` on `Option`
+  --> $DIR/map_flatten_fixable.rs:40:10
+   |
+LL |           .map(|_| {
+   |  __________^
+LL | | // we need some newlines
+LL | | // so that the span is big enough
+LL | | // for a splitted output of the diagnostic
+...  |
+LL | |         })
+LL | |         .flatten();
+   | |__________________^
+   |
+help: try replacing `map` with `and_then`
+   |
+LL ~         .and_then(|_| {
+LL + // we need some newlines
+LL + // so that the span is big enough
+   |
+help: and remove the `.flatten()`
+   |
+LL +             Some("")
+LL +  // whitespace beforehand is important as well
+LL ~         });
+   |
+
+error: aborting due to 8 previous errors
 

--- a/tests/ui/map_flatten_fixable.stderr
+++ b/tests/ui/map_flatten_fixable.stderr
@@ -77,7 +77,7 @@ LL |     let _: Result<_, &str> = (Ok(Ok(1))).and_then(|x| x);
    |                                          ~~~~~~~~~~~~~~~
 
 error: called `map(..).flatten()` on `Option`
-  --> $DIR/map_flatten_fixable.rs:40:10
+  --> $DIR/map_flatten_fixable.rs:59:10
    |
 LL |           .map(|_| {
    |  __________^

